File: Day10-19/19.c
#include <math.h>
#include <stdio.h>
#define MAX(a,b) ((a) > (b) ? a : b)


int elf_exchange(int n) {
    int b = 0;
    int t = n;
    while(t != 0) {
        b ++;
        t >>= 1;
    }
    n = n ^ (0x1 << (b-1));
    n = (n << 1) | 0x1;
    return n;
}

int elf_exchange_2(long n) {
    long l = (log2l(n) / log2l(3));
    long i = pow(3, l);
    return n == i ? n : MAX(n - i, 2*n - 3*i);
}

int main() {
    printf("Part 1: %d\n", elf_exchange(3005290));
    printf("Part 2: %d\n", elf_exchange_2(3005290));
}

File: Day10-19/19.py
from math import log


def elf_exchange(num):
    b = bin(num)[2:]
    return int(b[1:] + b[0], 2)


def elf_exchange_2(num):
    i = 3**int(log(num, 3))
    if num == i:
        return num
    else:
        return max(num - i, 2*num - 3*i)


if __name__ == '__main__':
    assert elf_exchange(5) == 3
    print("Part 1: %s" % elf_exchange(3005290))
    assert elf_exchange_2(5) == 2
    print("Part 2: %s" % elf_exchange_2(3005290))
